<a href="https://colab.research.google.com/github/vpste1/mlscrapbook/blob/master/PixelSimilarityAttempt.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [1]:
!pip install -Uqq fastbook
import fastbook
fastbook.setup_book()
from fastai.vision.all import *
from fastbook import *

matplotlib.rc('image', cmap='Greys')

     |████████████████████████████████| 727kB 8.1MB/s 
     |████████████████████████████████| 51kB 6.8MB/s 
     |████████████████████████████████| 1.1MB 20.1MB/s 
     |████████████████████████████████| 194kB 44.9MB/s 
     |████████████████████████████████| 61kB 9.7MB/s 
Mounted at /content/gdrive


In [2]:
path = untar_data(URLs.MNIST_SAMPLE)
Path.BASE_PATH = path

In [3]:
threes = (path/'train'/'3').ls().sorted()
sevens = (path/'train'/'7').ls().sorted()

Generate a list of images

In [4]:
seven_tensors = [tensor(Image.open(o)) for o in sevens]
three_tensors = [tensor(Image.open(o)) for o in threes]

Stack them into a single rank 3 tensor (3D matrix)

In [13]:
stacked_sevens = torch.stack(seven_tensors).float()/255
stacked_threes = torch.stack(three_tensors).float()/255
stacked_threes.shape

torch.Size([6131, 28, 28])

Generate a rank 2 tensor (2D matrix) of the mean across the 0th dimension. i.e. there are 6131 28x28 matrices; generate a single 28x28 matrix representing the averge across the 6131

In [14]:
mean3 = stacked_threes.mean(0)
mean7 = stacked_sevens.mean(0)

In [16]:
mean3.shape

torch.Size([28, 28])

Pick one of the "Threes" from the stack

In [ ]:
three =  stacked_threes[654]

Identify its "distance" from the mean in 2 different ways:
- Mean absolute (L1 norm) 
- RMS (L2 norm).

In [ ]:
dist_3_abs = (three - mean3).abs().mean()
dist_3_sqr = ((three - mean3)**2).abs().mean().sqrt()
dist_3_abs,dist_3_sqr


(tensor(0.1063), tensor(0.1868))

Do the same for the "Three" when comparing it to the mean across the 7s.

In [ ]:
dist_7_abs = (three - mean7).abs().mean()
dist_7_sqr = ((three - mean7)**2).abs().mean().sqrt()
dist_7_abs,dist_7_sqr

(tensor(0.1364), tensor(0.2694))

Pytorch provides the equivalent calculations as *loss functions* `l1_loss` and `mse_loss`. L1 norm (abs) is more lenient, whereas L2 norm (mse) will penalise bigger mistakes more heavily. 

In [ ]:
F.l1_loss(three.float(), mean3), F.mse_loss(three, mean3).sqrt()

(tensor(0.1063), tensor(0.1868))

In [ ]:
F.l1_loss(three.float(), mean7), F.mse_loss(three, mean7).sqrt()

(tensor(0.1364), tensor(0.2694))